lib/bn/fqp.ex

```diff
@@ -0,0 +1,308 @@
+defmodule BN.FQP do
+  defstruct [:coef, :modulus_coef, :dim]
+
+  alias BN.FQ
+
+  @type t :: %__MODULE__{
+          coef: [FQ.t()],
+          modulus_coef: [integer()]
+        }
+
+  @spec new([integer()], [integer()], keyword()) :: t()
+  def new(coef, modulus_coef, params \\ []) do
+    modulus = params[:modulus] || FQ.default_modulus()
+    coef_size = Enum.count(coef)
+    modulus_coef_size = Enum.count(modulus_coef)
+
+    if coef_size != modulus_coef_size,
+      do:
+        raise(ArgumentError,
+          message: "Coefficients and modulus coefficients have different dimensions"
+        )
+
+    fq_coef =
+      Enum.map(coef, fn coef_el ->
+        FQ.new(coef_el, modulus: modulus)
+      end)
+
+    %__MODULE__{
+      coef: fq_coef,
+      modulus_coef: modulus_coef,
+      dim: coef_size
+    }
+  end
+
+  @spec add(t(), t()) :: t()
```

Comment:
I think this function can raise an error, so the return type is not always `t()`.